Unify & redirect "How to create a bidder" things

rmloveland opened the pull request:
Specifically:

- Move 'How to add a Prebid 1.0 Bidder Adapter' content to 'How to Add a New Bidder Adapter'

- Copy edits to 'How to Add a New Bidder Adapter', including removing all of the "something something 1.0" language, since we are no longer accepting pre-1.0-format adapters

- Redirect 'How to add a Prebid 1.0 Bidder Adapter' to 'How to Add a New Bidder Adapter', and hide from nav

- Redirect 'How to add a new video adapter' to 'How to Add a New Bidder Adapter', and hide from nav

- Redirect 'Add video support to an existing adapter' to 'How to Add a New Bidder Adapter', and hide from nav

rmloveland commented:
Hey @mjacobsonny would appreciate any feedback on this change (details listed above).  Basically cleans up the adapter docs situation for a post-1.0 world.

mjacobsonny commented on dev-docs/add-video-support-to-an-existing-bidder-adaptor.md:
```diff
@@ -1,82 +1,10 @@
 ---
+redirect_to: "/dev-docs/bidder-adapter.html"
```
These redirects appear broken for me.  It looks like the page that exists is `/dev-docs/bidder-adaptor.html`.  Can we rename that page to `/dev-docs/bidder-adapter.html`?

rmloveland replied:
Addressed in d75cfe9

I fixed the busted redirects, and added a page `bidder-adapter` (with an 'e') so anyone using that spelling will land at the main page.

I don't want to rename the existing page because of the risk of breaking existing links/bookmarks and losing the current Google ranking for that URL.

mjacobsonny commented on dev-docs/bidder-adaptor.md:
```diff
@@ -12,259 +11,504 @@ nav_section: adapters
+* `modules/exampleBidAdapter.js`: the file containing the code for the adapter
+* `modules/exampleBidAdapter.md`: a markdown file containing key information about the adapter:
+   * The contact email of the adapter's maintainer.
+   * A test ad unit that will consistently return test creatives. This helps us to ensure future Prebid.js updates do not break your adapter.  Note that if your adapter supports video, outstream video, or native, you must also provide example parameters for each type.
```
Instead of "supports video, outstream video, or native", let's say something like "supports video (instream and/ or outstream context) or native..."

rmloveland replied:
Cool, updated in ac9f046

mjacobsonny commented on dev-docs/bidder-adaptor.md:
```diff
@@ -12,259 +11,504 @@ nav_section: adapters
+{% endhighlight %}
 
-In your PR to add the new adapter, please provide the following information:
+### Required adapter conventions
```
This section seems particularly crucial.  I wonder if we should break this out into it's own heading and move it up?

rmloveland replied:
Agreed, moved up to be the first subsection under 'Planning your Adapter' in 838ac64

mjacobsonny commented on dev-docs/bidder-adaptor.md:
```diff
@@ -12,259 +11,504 @@ nav_section: adapters
-{% endhighlight %}
+The AdUnit's `bid.params` object will define the parameters of your ad request.  You can include tag ID, site ID, ad size, keywords, and other data, such as video parameters.
+
+For more information about the kinds of information that can be passed using these parameters, see [the existing bidder parameters]({{site.baseurl}}/dev-docs/bidders.html).
```
Wording seems a bit awkward.

rmloveland replied:
Agreed, esp. that first sentence.  Tweaked in 838ac64

mjacobsonny commented on dev-docs/bidder-adaptor.md:
```diff
@@ -12,259 +11,504 @@ nav_section: adapters
-adaptermanager.registerBidAdapter(new XYZBidAdapter, 'xyz');
-module.exports = xyzBidAdapter;
-{% endhighlight %}
+The AdUnit's `bid.params` object will define the parameters of your ad request.  You can include tag ID, site ID, ad size, keywords, and other data, such as video parameters.
```
"Parameters specific to your adapter will be passed via the AdUnit's `bid.params` object."  Not sure we need to provide the specific "tag ID", "site ID", etc. examples here.

rmloveland replied:
Tweaked it slightly to just say "tag info" in 1c04d73.  Agree that tag ID/site ID aren't technically necessary, but I think it's good to provide some examples/color around what the parameters are generally used for before we link to the examples.

mjacobsonny commented on dev-docs/bidder-adaptor.md:
```diff
@@ -12,259 +11,504 @@ nav_section: adapters
 
-// the bidder API's ad response unit that has info like CPM, creative content
-var adUnit;
+See the [AppNexus AST adapter]({{site.baseurl}}/dev-docs/bidders.html#appnexusAst) for an example of how
```
I think we need to explain this instead of just pointing to the `appnexusAst` adapter.

Also, the `appnexusAst` adapter will not exist in Prebid 1.0.

rmloveland replied:
Good catch, updated this in 973c145 with the below - I tried to extrapolate from our AST adapter's design, let me know if you think it needs tweaking.

<blockquote>

Video parameters are often passed in from the ad unit in a `video` object.

The design of these parameters may vary depending on what your server-side bidder accepts.  If possible, we recommend using the video parameters in the [OpenRTB specification](https://iabtechlab.com/specifications-guidelines/openrtb/).

For examples of video parameters accepted by different adapters, see [the list of bidders with video demand]({{site.baseurl}}/dev-docs/bidders.html#bidder-video-native).

</blockquote>

mjacobsonny commented on dev-docs/bidder-adaptor.md:
```diff
@@ -12,259 +11,504 @@ nav_section: adapters
-bidObject.ad = adUnit.get('html');
-bidObject.width = adUnit.get('width');
-bidObject.height = adUnit.get('height');
+### Step 3: Respond with VAST or a VAST URL
```
"Respond with a VAST URL or raw VAST XML".

rmloveland replied:
Thanks, addressed in e36c2e1

mjacobsonny commented on dev-docs/bidder-adaptor.md:
```diff
@@ -12,259 +11,504 @@ nav_section: adapters
 
-// send the bidResponse object to bid manager with the adUnitCode.
-bidmanager.addBidResponse(adUnitCode, bidObject);
+When the bidder returns VAST or a VAST URL in its bid response, it needs to add the result into either `bid.vastXml` or `bid.vastUrl`. For example, here is some [code from the Tremor adapter](https://github.com/prebid/Prebid.js/blob/master/modules/tremorBidAdapter.js#L142) showing how it's done:
```
"The returned VAST URL or raw VAST XML should be added into `bid.vastUrl` or `bid.vastXml`, respectively."

rmloveland replied:
Also addressed in e36c2e1

mjacobsonny commented on dev-docs/bidder-adaptor.md:
```diff
@@ -12,259 +11,504 @@ nav_section: adapters
 {% endhighlight %}
 
-**`adUnitCode` in `addBidResponse`**
+#### If you would like to participate in auctions for outstream video ads
```
Support for the outstream video context should either be its own section under "Supporting Video" or we should have a separate step/ section of step 2 which describes properly ingesting the video context from the request.

rmloveland replied:
Took your suggestion to move this info into subsections of Step 2 in 5bf5a29 ('Ingesting the Video Context', and 'Outstream Video Renderers').  The content is basically the same as before, just reorganized into those sections.  Let me know if you have any additional feedback.

mjacobsonny commented on dev-docs/bidder-adaptor.md:
```diff
@@ -12,259 +11,504 @@ nav_section: adapters
-{% endhighlight %}
+In order for your bidder to support the native media type, you need 2 things:
+
+1. Your (server-side) bidder needs to respond with a bid that has native information.
```
Instead of "native information", we should always refer to "native assets".  We should also note that native bidders must be capable of ingesting the required and optional native assets specified on the `adUnit.mediaTypes.native` object.

rmloveland replied:
Addressed in a7a422e, specifically:

- Always refer to native *assets*

- Draw a clearer line between the server-side bidder's response and the (locally running) bidder adapter's response to Prebid.js

- Add a note that the adapter must be capable of ingesting native assets on `adUnit.mediaTypes.native`, and link to 'Show Native Ads' page for more info

mjacobsonny commented on dev-docs/bidder-adaptor.md:
```diff
@@ -12,259 +11,504 @@ nav_section: adapters
+For example, below is the [code in the AppNexus AST adapter](https://github.com/prebid/Prebid.js/blob/master/modules/appnexusAstBidAdapter.js#L192) that achieves #2.  As you can see, we:
+
+1. Check for native information on the bid response.
+2. Fill in the bid's `native` object with information from the bid.
```
It's confusing to use "bid" twice with different contexts.

rmloveland replied:
Also addressed in a7a422e

mjacobsonny commented on dev-docs/bidder-adaptor.md:
```diff
@@ -12,259 +11,504 @@ nav_section: adapters
+## Unit Tests
 
-## Helper functions
+Every adapter submission must include unit tests. See existing test suites in [test/spec/modules](https://github.com/prebid/Prebid.js/tree/master/test/spec/modules).
```
Are there any specific requirements for unit tests that we should call out here?

rmloveland replied:
Good callout - in e80533f I added a link to the Prebid.js CONTRIBUTING.md which has a bunch of testing requirements listed.

rmloveland commented on dev-docs/bidder-adaptor.md:
```diff
@@ -12,259 +11,504 @@ nav_section: adapters
+* `isBidRequestValid` - Verify the the `AdUnits.bids`, respond with `true` (valid) or `false` (invalid).
+* `buildRequests` - Takes an array of valid bid requests, all of which are guaranteed to have passed the `isBidRequestValid()` test.
+* `interpretResponse` - Parse the response and generate one or more bid objects.
+* `getUserSyncs` - If the publisher allows user-sync activity, the platform will call this function and the adapter may register pixels and/or iframe user syncs.
```
Good call - addressed in e3ddc14

rmloveland commented on dev-docs/bidder-adaptor.md:
```diff
@@ -12,259 +11,504 @@ nav_section: adapters
+        currency: CURRENCY,
+        netRevenue: true,
+        ttl: TIME_TO_LIVE,
+        referrer: REFERER,
```
Looks like that snuck in during the original ["Prebid 1.0 bidder adapter" PR](https://github.com/prebid/prebid.github.io/pull/343/files) - removed in 61a3259

rmloveland commented on dev-docs/bidder-adaptor.md:
```diff
@@ -12,259 +11,504 @@ nav_section: adapters
+| `vastXml`    | Either this or `vastUrl` required for video | XML for VAST document to be cached for later retrieval.                                                                                       | `<VAST version="3.0">...`            |
+| `dealId`     | Optional                                    | Deal ID                                                                                                                                       | `"123abc"`                           |
+
+### Register User Syncs
```
Updated in e2a0ca2 to focus on just one method, `getUserSyncs` callback, after some discussion with engineering team.

Included documentation on the method and links to examples in the Prebid.js repo.

rmloveland commented on dev-docs/bidder-adaptor.md:
```diff
@@ -12,259 +11,504 @@ nav_section: adapters
+{: .table .table-bordered .table-striped }
+| Attribute | Type             | Description                                                        | Example Value             |
+|-----------+------------------+--------------------------------------------------------------------+---------------------------|
+| type      | string           | Which HTTP method should be used                                   | GET/POST                  |
```
Yes.  I read the code for a few adapters to verify.  Guessing the table was written before the design was later finalized.   Updated in b5876f3

rmloveland commented:
I think I've addressed all the feedback, @mjacobsonny  - hoping to merge ASAP, let me know if you would like any additional changes.  Thanks for the thorough review!
